diff --git a/genl/Makefile b/genl/Makefile
index 2b7a45b..3d66a97 100644
--- a/genl/Makefile
+++ b/genl/Makefile
@@ -14,20 +14,21 @@ GENLOBJ += $(GENLMODULES)
 GENLLIB :=
 
 ifeq ($(SHARED_LIBS),y)
-LDFLAGS += -Wl,-export-dynamic
+LDFLAGS += -Wl,--dynamic-list=dynsyms.list
 LDLIBS  += -lm -ldl
 endif
 
 all: genl
 
 genl: $(GENLOBJ) $(LIBNETLINK) $(LIBUTIL) $(GENLLIB)
-	$(QUIET_LINK)$(CC) $^ $(LDFLAGS) $(LDLIBS) -o $@
+	$(QUIET_LINK)$(CC) $(filter-out dynsyms.list, $^) $(LDFLAGS) $(LDLIBS) -o $@
 
 install: all
 	install -m 0755 genl $(DESTDIR)$(SBINDIR)
 
 clean:
 	rm -f $(GENLOBJ) $(GENLLIB) genl
+	rm -f dynsyms.list
 
 ifneq ($(SHARED_LIBS),y)
 
@@ -39,4 +40,15 @@ static-syms.h: $(wildcard *.c)
 		sed -n '/'$$s'[^ ]* =/{s:.* \([^ ]*'$$s'[^ ]*\) .*:extern char \1[] __attribute__((weak)); if (!strcmp(sym, "\1")) return \1;:;p}' $$files ; \
 	done > $@
 
+else
+
+genl: dynsyms.list
+dynsyms.list: $(wildcard *.c)
+	files="$^" ; \
+	echo "{" > $@ ; \
+	for s in `grep -B 3 '\<dlsym' $$files | sed -n '/snprintf/{s:.*"\([^"]*\)".*:\1:;s:%s::;p}'` ; do \
+		sed -n '/'$$s'[^ ]* =/{s:.* \([^ ]*'$$s'[^ ]*\) .*:\1;:;p}' $$files ; \
+	done >> $@ ; \
+	echo "};" >> $@
+
 endif
diff --git a/ip/Makefile b/ip/Makefile
index a88f936..6182325 100644
--- a/ip/Makefile
+++ b/ip/Makefile
@@ -23,7 +23,7 @@ TARGETS=ip rtmon
 all: $(TARGETS) $(SCRIPTS)
 
 ip: $(IPOBJ) $(LIBNETLINK)
-	$(QUIET_LINK)$(CC) $^ $(LDFLAGS) $(LDLIBS) -o $@
+	$(QUIET_LINK)$(CC) $(filter-out dynsyms.list, $^) $(LDFLAGS) $(LDLIBS) -o $@
 
 rtmon: $(RTMONOBJ)
 	$(QUIET_LINK)$(CC) $^ $(LDFLAGS) $(LDLIBS) -o $@
@@ -34,12 +34,22 @@ install: all
 
 clean:
 	rm -f $(ALLOBJ) $(TARGETS)
+	rm -f dynsyms.list
 
 SHARED_LIBS ?= y
 ifeq ($(SHARED_LIBS),y)
 
 LDLIBS += -ldl
-LDFLAGS += -Wl,-export-dynamic
+LDFLAGS += -Wl,--dynamic-list=dynsyms.list
+
+ip: dynsyms.list
+dynsyms.list: $(STATIC_SYM_SOURCES)
+	files="$^" ; \
+	echo "{" > $@ ; \
+	for s in `grep -B 3 '\<dlsym' $$files | sed -n '/snprintf/{s:.*"\([^"]*\)".*:\1:;s:%s::;p}'` ; do \
+		sed -n '/'$$s'[^ ]* =/{s:.* \([^ ]*'$$s'[^ ]*\) .*:\1;:;p}' $$files ; \
+	done >> $@ ; \
+	echo "};" >> $@
 
 else
 
diff --git a/tc/Makefile b/tc/Makefile
index 2edaf2c..d7d91fa 100644
--- a/tc/Makefile
+++ b/tc/Makefile
@@ -107,7 +107,7 @@ LDLIBS += -L. -lm
 
 ifeq ($(SHARED_LIBS),y)
 LDLIBS += -ldl
-LDFLAGS += -Wl,-export-dynamic
+LDFLAGS += -Wl,--dynamic-list=dynsyms.list
 endif
 
 TCLIB := tc_core.o
@@ -134,7 +134,7 @@ MODDESTDIR := $(DESTDIR)$(LIBDIR)/tc
 all: tc $(TCSO)
 
 tc: $(TCOBJ) $(LIBNETLINK) libtc.a
-	$(QUIET_LINK)$(CC) $^ $(LDFLAGS) $(LDLIBS) -o $@
+	$(QUIET_LINK)$(CC) $(filter-out dynsyms.list, $^) $(LDFLAGS) $(LDLIBS) -o $@
 
 libtc.a: $(TCLIB)
 	$(QUIET_AR)$(AR) rcs $@ $^
@@ -156,6 +156,7 @@ install: all
 clean:
 	rm -f $(TCOBJ) $(TCLIB) libtc.a tc *.so emp_ematch.yacc.h; \
 	rm -f emp_ematch.yacc.*
+	rm -f dynsyms.list
 
 q_atm.so: q_atm.c
 	$(QUIET_CC)$(CC) $(CFLAGS) $(CPPFLAGS) $(LDFLAGS) -shared -fpic -o q_atm.so q_atm.c -latm
@@ -195,4 +196,15 @@ static-syms.h: $(wildcard *.c)
 		sed -n '/'$$s'[^ ]* =/{s:.* \([^ ]*'$$s'[^ ]*\) .*:extern char \1[] __attribute__((weak)); if (!strcmp(sym, "\1")) return \1;:;p}' $$files ; \
 	done > $@
 
+else
+
+tc: dynsyms.list
+dynsyms.list: $(wildcard *.c)
+	files="$(filter-out $(patsubst %.so,%.c,$(TCSO)), $^)" ; \
+	echo "{" > $@ ; \
+	for s in `grep -B 3 '\<dlsym' $$files | sed -n '/snprintf/{s:.*"\([^"]*\)".*:\1:;s:%s::;p}'` ; do \
+		sed -n '/'$$s'[^ ]* =/{s:.* \([^ ]*'$$s'[^ ]*\) .*:\1;:;p}' $$files ; \
+	done >> $@ ; \
+	echo "show_stats; print_tm; parse_rtattr; };" >> $@
+
 endif
